Auth sub

SwaySway opened the pull request:
*Issue #, if available:*
re #1766 #1043

*Description of changes:*
adding protection on subscriptions

BREAKING CHANGE: If an owner is included in the auth directive it will either be a requirement if it's the only rule or an optional input if used with other rules

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

SwaySway commented on packages/graphql-auth-transformer/src/resources.ts:
```diff
@@ -276,6 +275,69 @@ groupsField: "${rule.groupsField || DEFAULT_GROUPS_FIELD}" }`
+        for (const rule of rules) {
+            const ownerAttribute = rule.ownerField || DEFAULT_OWNER_FIELD
+            const rawUsername = rule.identityField || DEFAULT_IDENTITY_FIELD
+            const isUsern = isUsername(rawUsername)
```
using the abbreviation for isUsern

attilah replied:
Please do not.

SwaySway replied:
Changed the naming convention

SwaySway commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -172,6 +176,7 @@ export class ModelAuthTransformer extends Transformer {
         if (!modelDirective) {
             throw new InvalidDirectiveError('Types annotated with @auth must also be annotated with @model.')
         }
+        MODEL_DIRECTIVE_STACK = def.name.value;
```
good point - I can refer to the model directive stack name with def.name.value. So I've made this change in the code where it no longer has a need to ref modelDirectiveStack

SwaySway commented on packages/graphql-auth-transformer/src/resources.ts:
```diff
@@ -689,4 +768,34 @@ identityField: "${rule.identityField || DEFAULT_IDENTITY_FIELD}" }`
+        });
+    }
+
+    public subscriptionHas(subscription: any): Object {
```
Changed in the new revision to getDirectiveArgs()

SwaySway commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -187,6 +192,31 @@ export class ModelAuthTransformer extends Transformer {
+
+        // Check if subscriptions is enabled
+        const subscription = modelDirective.arguments.find((arg) => arg.name.value === 'subscriptions');
+        if (subscription && subscription.value.kind !== "NullValue") {
```
Changed in the new revision

SwaySway commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -872,6 +902,191 @@ All @auth directives used on field definitions are performed when the field is r
         }
     }
 
+    private protectCreateSubcriptions(ctx: TransformerContext, rules: AuthRule[],  parent: ObjectTypeDefinitionNode, nameOverride?: string) {
```
changed in the new revision

SwaySway commented on packages/graphql-auth-transformer/src/resources.ts:
```diff
@@ -689,4 +768,34 @@ identityField: "${rule.identityField || DEFAULT_IDENTITY_FIELD}" }`
+    public generateSubscriptionResolver(type: string, fieldName: string, subscriptionTypeName: string = 'Subscription') {
+        return new AppSync.Resolver({
+            ApiId: Fn.GetAtt(ResourceConstants.RESOURCES.GraphQLAPILogicalID, 'ApiId'),
+            DataSourceName: LOCAL_DATASOURCE,
```
Added the nonedatasource for the subscription resolvers

yuth commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -872,6 +1012,128 @@ All @auth directives used on field definitions are performed when the field is r
+        const resolver = this.resources.generateSubscriptionResolver(fieldName);
+
+        // creates the none data source
+        const noneDS = ctx.getResource(ResourceConstants.RESOURCES.NoneDataSource)
```
Do we want to create a data source if its not being used? Should you set this only if there is `rules.length > 0`

SwaySway replied:
Creating a none data source if the subscription resolvers are being created.

yuth commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -478,10 +602,22 @@ Static group authorization should perform as expected.`
     }
 
     private validateFieldRules(rules: AuthRule[]) {
+        // check if the rules in the field are a subset of the rules in authrule
```
Is this relevant anymore

SwaySway replied:
Removed Comment

yuth commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -537,15 +673,19 @@ All @auth directives used on field definitions are performed when the field is r
             throwIfUnauthorizedExpression
-        ])
+        ]
+        expressions.push(
```
Do you need this?

SwaySway replied:
That was from a prev change removed in latest commit

yuth commented on packages/graphql-dynamodb-transformer/src/DynamoDBModelTransformer.ts:
```diff
@@ -119,6 +132,12 @@ export class DynamoDBModelTransformer extends Transformer {
             }
         )
 
+        // check if auth if enabled
```
cruft?

yuth commented on packages/graphql-transformer-core/src/util/transformConfig.ts:
```diff
@@ -50,6 +50,12 @@ export interface TransformConfig {
      */
     Migration?: TransformMigrationConfig;
+
+    /**
```
Change the comment since the flag has been changed to version number

SwaySway replied:
Changed comment

RossWilliams commented on packages/graphql-auth-transformer/src/resources.ts:
```diff
@@ -162,8 +171,35 @@ export class ResourceFactory {
             ])
         ])
+
+        const customGroupClaim: Expression = block('Static Group Authorization Checks with Custom Groups', [
```
There is a lot of code duplication here that could be avoided. You are doing three extra steps: adding a comment when a custom claim is used, calling a different method to set userGroups, and handling the claim field being either a string or an array. You could remove this duplication (and source of future mis-alignment / bugs) by updating setCustomClaim function to convert any string claim to an array. Further, if you change setUserGroups to take a claim name as the first parameter with a default value of 'cognito:groups' you can completely remove this method, and remove the setCustomClaim method.

SwaySway replied:
Good call out - I've edited setUserGroups to allow an optional param for a custom group claim.

RossWilliams commented on packages/graphql-auth-transformer/src/resources.ts:
```diff
@@ -685,4 +828,43 @@ identityField: "${rule.identityField || DEFAULT_IDENTITY_FIELD}" }`
     }
-}
+
+    public setCustomClaim(customGroups: string): SetNode {
```
This method is problematic as you are setting userGroups to an unknown type, while setUserGroups uses cognito:groups which is known to be an array type. userGroups should have a stable type. Ive left a comment above on how this method can be removed completely.

SwaySway replied:
Added this check in setUsergroups where it'll reassign the the string into an array

RossWilliams commented on packages/graphql-auth-transformer/src/resources.ts:
```diff
@@ -217,7 +253,15 @@ groupsField: "${rule.groupsField || DEFAULT_GROUPS_FIELD}" }`
         for (const rule of rules) {
+            if (rule.groupClaim) {
+                if (customClaim) {
+                    throw new InvalidDirectiveError('@auth directive currently only supports one source for groupClaim!')
```
This restriction isn't necessary if you set the userGroups as the first item (or after the comments) of the groupAuthorizationExpressions.
```typescript
groupAuthorizationExpressions = groupAuthorizationExpressions.concat(
  formatComment ?
    comment(formatComment(rule)) :
    comment(`Authorization rule: { allow: ${rule.allow}, groupsField: "${groupsAttribute}" }`),
  this.setUserGroups(rule.groupClaim),
  set(
     ref(variableToSet),
    raw(`$util.defaultIfNull($${variableToSet}, false)`)
  ),
```

assumes you adopt the setUserGroups overload, otherwise you can use a ternary to choose the right function to call.

SwaySway replied:
Ah I see each rule would have it's own respective claim. Though I'm not sure of the use case since the groups could exist in the same groupClaim. This might warrant a separate pr where we can discuss a use case for this.

RossWilliams replied:
I already use this on my Amplify fork, and do have a use case. I have an auth rule based on the TenantId claim, and another auth rule based on a TeamId, and another AuthId based on the AppGroupId.
Only requires moving the setUserGroups to inside the forEach loop, so each rules evaluates against its own claimGroup. And then you can delete ~15 lines of code.

undefobj replied:
@RossWilliams it's less about the code and more related to complexity. We're already adding new functionality here which is pretty flexible, and adding multiple claims that could overlap permissions is complexity on already a complex concept which could have downstream impact to the authorization rules which needs evaluation. At this time leaving this as-is seems to be a more prudent choice which we could re-evaluate as a separate feature in the future.

RossWilliams commented on packages/graphql-auth-transformer/src/resources.ts:
```diff
@@ -405,7 +520,14 @@ identityField: "${rule.identityField || DEFAULT_IDENTITY_FIELD}" }`
         for (const rule of rules) {
+            if (rule.groupClaim) {
+                if (customClaim) {
+                    throw new InvalidDirectiveError('@auth directive currently only supports one source for groupClaim!')
```
See above solution to remove this restriction on the create operation.

RossWilliams commented on packages/graphql-auth-transformer/src/resources.ts:
```diff
@@ -504,7 +627,14 @@ identityField: "${rule.identityField || DEFAULT_IDENTITY_FIELD}" }`
         for (const rule of rules) {
+            if (rule.groupClaim) {
+                if (customClaim) {
+                    throw new InvalidDirectiveError('@auth directive currently only supports one source for groupClaim!')
```
See above comment to remove this restriction.

RossWilliams commented on packages/graphql-auth-transformer/src/resources.ts:
```diff
@@ -138,21 +139,29 @@ export class ResourceFactory {
+                customClaim = rule.groupClaim;
+            }
         }
         // TODO: Enhance cognito:groups to work with non cognito based auth.
```
Comment can be removed with this PR

RossWilliams commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -267,6 +352,13 @@ Static group authorization should perform as expected.`
                 resolver = this.resources.blankResolver(typeName, fieldName)
             }
             const authExpression = this.authorizationExpressionOnSingleObject(rules, 'ctx.source')
+            if (protectPrivateFields) {
```
The modified tests suggest this isn't correct.
context.result.operation is set to query in setOperationExpression, but then context.source.operation is checked for query in operationCheckExpression, fails, and returns null.

RossWilliams commented on packages/graphql-transformers-e2e-tests/src/__tests__/ConnectionsWithAuthTests.e2e.test.ts:
```diff
@@ -276,7 +276,7 @@ test('Testing reading an owner protected field as a non owner', async () => {
     expect(response1.data.createFieldProtected.id).toEqual("1")
     expect(response1.data.createFieldProtected.owner).toEqual(USERNAME1)
-    expect(response1.data.createFieldProtected.ownerOnly).toEqual("owner-protected")
+    expect(response1.data.createFieldProtected.ownerOnly).toBeNull()
```
I don't understand why this test should be updated. The field is protected to only allow the owner, but USERNAME1 is creating the object and USERNAME1 is set as the owner field. ownerOnly should not be null for the owner. Possible source of issue left as comment above in the protectPrivateFields check block.

RossWilliams replied:
@yuth You may want to review these test changes as you have marked this PR as approved.

SwaySway replied:
The behavior of per field auth has changed - the PR for docs elaborates on this more
[Docs PR #868 L941](https://github.com/aws-amplify/docs/pull/868/files#diff-8076c2db248c895d2ade586451fa9b55R944#L941)

RossWilliams commented on packages/graphql-transformers-e2e-tests/src/__tests__/ConnectionsWithAuthTests.e2e.test.ts:
```diff
@@ -372,7 +372,7 @@ test('Test that owners cannot set the field of a FieldProtected object unless au
     expect(response1.data.createFieldProtected.id).toEqual("2")
     expect(response1.data.createFieldProtected.owner).toEqual(USERNAME1)
-    expect(response1.data.createFieldProtected.ownerOnly).toEqual("owner-protected")
+    expect(response1.data.createFieldProtected.ownerOnly).toBeNull()
```
This test should not have been updated.

RossWilliams commented on packages/graphql-transformers-e2e-tests/src/__tests__/ConnectionsWithAuthTests.e2e.test.ts:
```diff
@@ -414,7 +414,7 @@ test('Test that owners cannot update the field of a FieldProtected object unless
     expect(response1.data.createFieldProtected.id).not.toBeNull()
     expect(response1.data.createFieldProtected.owner).toEqual(USERNAME1)
-    expect(response1.data.createFieldProtected.ownerOnly).toEqual("owner-protected")
+    expect(response1.data.createFieldProtected.ownerOnly).toBeNull()
```
This test should not have been updated. See above comment.

RossWilliams commented on packages/graphql-transformers-e2e-tests/src/__tests__/ConnectionsWithAuthTests.e2e.test.ts:
```diff
@@ -439,7 +439,7 @@ test('Test that owners cannot update the field of a FieldProtected object unless
     expect(response3.data.updateFieldProtected.id).toEqual(response1.data.createFieldProtected.id)
     expect(response3.data.updateFieldProtected.owner).toEqual(USERNAME1)
-    expect(response3.data.updateFieldProtected.ownerOnly).toEqual("updated")
+    expect(response3.data.updateFieldProtected.ownerOnly).toBeNull()
```
This test does not have a reason to change. See above comment.

undefobj commented:
@SwaySway per chat we need to respond with `null` and not `{ "ok": true }` with an authorized request on subscriptions otherwise the GraphQL schema will give a type error and fail the subscription request from a client

undefobj commented:
Also note that for existing projects the check:

```
? The default behavoir for @auth has changed in the latest version of Amplify
Read here for details: https://aws-amplify.github.io/docs/cli-toolchain/graphql#authorizing-subscriptions 
 Do you wish to continue? Yes
authsubstestt $
```

This is not working correctly, it always will do a `process.exit()` which blocks an `amplify push` no matter what you select. We need to look at this logic and see what the bug is. @SwaySway @kaustavghosh06

SwaySway commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -221,6 +301,8 @@ Static group authorization should perform as expected.`
         // Get and validate the auth rules.
         const rules = getArg('rules', []) as AuthRule[]
         this.validateFieldRules(rules)
+        // if the object has auth enabled and the field rules are a superset of the object rules
```
Removed Comment

SwaySway commented on packages/amplify-provider-awscloudformation/lib/transform-graphql-schema.js:
```diff
@@ -22,11 +26,55 @@ const schemaDirName = 'schema';
+}
+
+/**
+ * @TODO change authWarning to version number to compare
```
Changed @TODO to add a version mapping

SwaySway commented:
> @SwaySway per chat we need to respond with `null` and not `{ "ok": true }` with an authorized request on subscriptions otherwise the GraphQL schema will give a type error and fail the subscription request from a client

I've added a change on the subscription resolver to return null
> This is not working correctly, it always will do a `process.exit()` which blocks an `amplify push` no matter what you select. We need to look at this logic and see what the bug is. @SwaySway @kaustavghosh06

Corrected in the latest commit

SwaySway commented:
Added an e2e test in `PerFieldAuthTests.e2e.test.ts` to confirm new `@auth` behavior

attilah commented on packages/amplify-provider-awscloudformation/lib/transform-graphql-schema.js:
```diff
@@ -22,11 +26,61 @@ const schemaDirName = 'schema';
+    throw new Error(`You are trying to
+use the @auth directive without enabling Amazon Cognito user pools for your API.
+Run \`amplify update api\` and choose
+"Amazon Cognito User Pool" as the authorization type for the API.`);
```
In the light of multi-auth being merged, we should rephrase this to 'enable' or 'add' Cognito, because this wording could suggest that it works if *only* Cognito enabled.

attilah commented on packages/amplify-provider-awscloudformation/lib/transform-graphql-schema.js:
```diff
@@ -22,11 +26,61 @@ const schemaDirName = 'schema';
+      const response = await inquirer.prompt({
+        name: 'transformerConfig',
+        type: 'confirm',
+        message: `${versionChangeMessage} \n Do you wish to continue?`,
```
We're printing this message anyway, but in one case as part of a prompt, otherwise as a warning, why not print it as warning anyway before the if, then have the prompt if needed?

nit: I'd not add newlines to messages as prefix/suffix, since once we'll add logging it will be hard to cleanup all these 'additions'.

attilah commented on packages/amplify-provider-awscloudformation/lib/transform-graphql-schema.js:
```diff
@@ -22,11 +26,61 @@ const schemaDirName = 'schema';
+        default: false,
+      });
+      if (!response.transformerConfig) {
+        process.exit(0);
```
Why is the hard process exit, can't we throw and let SEH handle the exit from the process?

attilah commented on packages/amplify-provider-awscloudformation/lib/transform-graphql-schema.js:
```diff
@@ -22,11 +26,61 @@ const schemaDirName = 'schema';
+      }
+    }
+  }
+  transformerConfig.Version = 3.0;
```
3.0 as number? I think for versions in general we'd need something that can be parsed AND compared with ```semver``` lib to make differentiation later much easier.

attilah commented on packages/graphql-auth-transformer/package.json:
```diff
@@ -7,7 +7,8 @@
     "test-ci": "jest --ci -i",
     "build": "tsc",
-    "clean": "rm -rf ./lib"
+    "clean": "rm -rf ./lib",
```
nit: rm is *nix only, we should use rimraf everywhere, since it will break on windows. Yes I know we're currently breaking dev on windows, but it would be nice to no to add new 'problems'.

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -88,12 +121,16 @@ export class ModelAuthTransformer extends Transformer {
                 # Specifies the auth rule's strategy. Allowed values are 'owner' and 'groups'.
                 allow: AuthStrategy!
 
+                # Legacy name for identifyClaim
```
typo: identiFyClaim, F instead of T multiple times.

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -1,26 +1,58 @@
+    delete?: string;
+}
+
+enum ModelSubscriptionLevel {
```
Why assigning string literals to enum value? if we need string, then use type like for auth modes. That gives us value safety with typescript.

Why the ```ON``` value is not represented in the enum?

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -103,6 +140,10 @@ export class ModelAuthTransformer extends Transformer {
                 # Specifies the field of type String or [String] that contains group(s) that can access the object.
                 groupsField: String
```
Why don't we deprecate ```groupsField``` isn't it the same what we do with ```identityField``` ?

attilah commented on packages/graphql-auth-transformer/src/AuthRule.ts:
```diff
@@ -5,9 +5,11 @@ export interface AuthRule {
     operations?: ModelOperation[]
     queries?: ModelQuery[]
     mutations?: ModelMutation[]
+    groupClaim?: string;
```
Please move it under ```groupsField``` for maintainability.

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -1,26 +1,58 @@
+}
+
+interface SubscriptionNameMap {
+    onCreate?: string[];
```
With this structure you cannot configure level per operation. You can't make ```onCreate``` ```PUBLIC``` and protect only ```onUpdate``` and ```onDelete```. Are we ok with it?

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -187,6 +228,34 @@ export class ModelAuthTransformer extends Transformer {
+        if ("subscriptions" in directiveArguments) {
+            const subscription = directiveArguments.subscriptions;
+            // check that subscription is not set to off
+            if (subscription && !subscription.level) {
```
Should not we validate the value of ```level``` to ```OFF``` here?

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -187,6 +228,34 @@ export class ModelAuthTransformer extends Transformer {
+                        this.protectOnDeleteSubcription(ctx, operationRules.delete, def, fieldName);
+                    })
+                }
+            } else { return; }
```
nit: break into multiple lines for readability.

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -205,7 +274,17 @@ export class ModelAuthTransformer extends Transformer {
+        const modelDirective = parent.directives.find((dir) => dir.name.value === 'model')
+        const parentModelArgs: ModelDirectiveArgs = modelDirective ? getDirectiveArguments(modelDirective) : {};
+        //  check if subscriptions is explicity disabled or if it's set to public
+        if ('subscriptions' in parentModelArgs &&
```
The checks this way are not really explicit, based on a lot of JS behavior. What about making these checks more explicit? I can't see how are we checking for ```PUBLIC``` here? if subscriptions are explicitly disabled then it should have a ```null``` value.

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -236,24 +315,27 @@ Static group authorization should perform as expected.`
 
-    private protectField(ctx: TransformerContext, typeName: string, fieldName: string, rules: AuthRule[]) {
+    private protectField(ctx: TransformerContext, typeName: string,
+        field: FieldDefinitionNode, rules: AuthRule[], protectPrivateFields?: boolean) {
```
Why is protectPrivateFields optional? make it mandatory with a default value, then code would be more explicit inside IMHO.

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -264,9 +346,20 @@ Static group authorization should perform as expected.`
+Either make the field optional, set auth on the object and not the field, or disable subscriptions for the object (setting level to OFF or PUBLIC)\n`)
+                }
+                // add operation to queryField
+                this.protectMutations(ctx, typeName, 'mutation')
```
Don't use ```'mutation'``` literal here, use ```ctx.getMutationTypeName```. What other value this parameter can have? Because the called function name suggests only mutations are supported. The comment above suggests queryField, which is not correct.

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -478,6 +597,17 @@ Static group authorization should perform as expected.`
             }
+            if ( allow === 'groups' && (identityClaim || identityField)) {
+                throw new InvalidDirectiveError(`
+                @auth identifyField/Claim is only for 'allow: owner'`)
```
Friendlier message: "can only be used" or something similar

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -478,6 +597,17 @@ Static group authorization should perform as expected.`
+            }
+            if (allow === 'owner' && groupClaim) {
+                throw new InvalidDirectiveError(`
+                @auth groupClaim is only for 'allow: groups'`);
```
Same as above.

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -478,6 +597,17 @@ Static group authorization should perform as expected.`
+                throw new InvalidDirectiveError(`
+                @auth groupClaim is only for 'allow: groups'`);
+            }
+            if ( groupsField && groups) {
```
I think this validation is incorrect, since ```groups``` defines the list of allowed groups, ```groupsField``` specifies which field to look for the group values in the type.

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -469,7 +586,9 @@ Static group authorization should perform as expected.`
     private validateRules(rules: AuthRule[]) {
         for (const rule of rules) {
-            const { queries, mutations, operations } = rule;
+            const { queries, mutations, operations,
```
Missing validations:
- groupField and groupClaim is mutually exclusive
- identityField and identityClaim is mutually exclusive

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -872,6 +1001,128 @@ All @auth directives used on field definitions are performed when the field is r
     }
 
+    // OnCreate Subscription
+    private protectOnCreateSubcription(ctx: TransformerContext, rules: AuthRule[],
```
Why have 3 extra methods if the graphqlName could be just a parameter for ```protectSubscription```?

attilah commented on packages/graphql-auth-transformer/src/ModelAuthTransformer.ts:
```diff
@@ -872,6 +1001,128 @@ All @auth directives used on field definitions are performed when the field is r
+        if (hasOwner) {
+            this.addOwner(ctx, parent.name.value);
+        }
+        if (rules.length === 1 && hasOwner ) {
```
Code identation is wrong, some comments about why are we adding rules would be nice for future developers, consequent if could be an ```else if```?

attilah commented on packages/graphql-auth-transformer/src/resources.ts:
```diff
@@ -138,21 +139,29 @@ export class ResourceFactory {
             }
+            if (rule.groupClaim) {
+                if (customClaim) {
+                    throw new InvalidDirectiveError('@auth directive currently only supports one source for groupClaim!')
```
format: no need for exclamation mark, also we should output what type, what directive, what claim we found, what extra claim we found, that helps the developers to locate the error easier.

attilah commented on packages/graphql-dynamodb-transformer/src/DynamoDBModelTransformer.ts:
```diff
@@ -28,16 +28,22 @@ interface MutationNameMap {
     delete?: string;
 }
 
+enum ModelSubscriptionLevel {
```
Duplicate from ```AuthTransformer```, can't we share these types somehow? It is also missing the ```ON``` value, see my comment above at ```AuthTransformer```

attilah commented on packages/graphql-dynamodb-transformer/src/DynamoDBModelTransformer.ts:
```diff
@@ -79,6 +85,12 @@ export class DynamoDBModelTransformer extends Transformer {
                 onDelete: [String]
+                level: ModelSubscriptionLevel
+            }
+            enum ModelSubscriptionLevel {
```
We standardized on lowercase enum values for our directives, see ```AuthStrategy``` et al.

attilah commented on packages/graphql-transformer-core/src/util/sanity-check.ts:
```diff
@@ -261,6 +262,10 @@ export function cantHaveMoreThan200Resources(diffs: Diff[], currentBuild: Diffab
 }
 
+export function authSubscriptions(diff: Diff[]) {
+    // check if the subscription
```
Unfinished function, TODO?
